log.go

```diff
@@ -252,20 +327,51 @@ func (l *raftLog) commitTo(tocommit uint64) {
 	}
 }
 
-func (l *raftLog) appliedTo(i uint64) {
-	if i == 0 {
-		return
-	}
+func (l *raftLog) appliedTo(i uint64, size entryEncodingSize) {
 	if l.committed < i || i < l.applied {
 		l.logger.Panicf("applied(%d) is out of range [prevApplied(%d), committed(%d)]", i, l.applied, l.committed)
 	}
 	l.applied = i
+	l.applying = max(l.applying, i)
+	if l.applyingEntsSize > size {
+		l.applyingEntsSize -= size
+	} else {
+		// Defense against underflow.
+		l.applyingEntsSize = 0
+	}
+	l.applyingEntsPaused = l.applyingEntsSize >= l.maxApplyingEntsSize
+}
+
+func (l *raftLog) acceptApplying(i uint64, size entryEncodingSize, allowUnstable bool) {
+	if l.committed < i {
+		l.logger.Panicf("applying(%d) is out of range [prevApplying(%d), committed(%d)]", i, l.applying, l.committed)
+	}
+	l.applying = i
+	l.applyingEntsSize += size
+	// Determine whether to pause entry application until some progress is
+	// acknowledged. We pause in two cases:
+	// 1. the outstanding entry size equals or exceeds the maximum size.
+	// 2. the outstanding entry size does not equal the maximum size, but we
+	//    determine that the next entry in the log will push us over the limit.
+	//    We determine this by comparing the last entry returned from
+	//    raftLog.nextCommittedEnts to the maximum entry that the method was
+	//    allowed to return had there been no size limit. If these indexes are
+	//    not equal, then the returned entries slice must have been truncated to
+	//    adhere to the memory limit.
+	l.applyingEntsPaused = l.applyingEntsSize >= l.maxApplyingEntsSize ||
+		i < l.maxAppliableIndex(allowUnstable)
```

Comment:
Does the second condition imply the first? In other words, can we remove the first part?

Reply:
Not quite. It is possible for `i == l.maxAppliableIndex(allowUnstable)` but for the first condition to be true in two different ways:
- `l.applyingEntsSize == l.maxApplyingEntsSize`: The appliable entries in the log were exactly `maxApplyingEntsSize`
- `l.applyingEntsSize >  l.maxApplyingEntsSize`: The first appliable entry in the log was larger than `maxApplyingEntsSize`

I spent some time trying to avoid needing to add both conditions, but I don't see how to eliminate either of them.